File: cpp/leetcode/editor/cn/[5]最长回文子串.cpp
#include <vector>
#include "util.h"
#include <map>
#include <string>
#include <utility>
#include <memory>

using namespace std;

//leetcode submit region begin(Prohibit modification and deletion)
class Solution {
public:
    string longestPalindrome(string s) {
        string_view res;
        for (auto left = s.begin(); left != s.end(); left++) {
            auto p1 = palindrome(s, left, left);
            if (p1.length() > res.length()) res = p1;

            auto p2 = palindrome(s, left, left + 1);
            if (p2.length() > res.length()) res = p2;
        }

        return string(res);
    }

    static string_view palindrome(const string &s, string::iterator left, string::iterator right) {
        while (left >= s.begin() && right <= s.end() && *left == *right) {
            left--;
            right++;
        }
        return {left + 1, right};
    }
};
//leetcode submit region end(Prohibit modification and deletion)


int main(int argv, char *argc[]) {
}

File: ts/leetcode/editor/cn/[5]最长回文子串.ts
//leetcode submit region begin(Prohibit modification and deletion)
function palindrome(s: string, left: number, right: number): string {
    if (left < 0 || right >= s.length || s[left] != s[right]) return s.substring(left + 1, right)
    else return palindrome(s, left - 1, right + 1)
}

function longestPalindrome(s: string): string {
    let res = ""
    for (let i = 0; i < s.length; i++) {
        const l1 = palindrome(s, i, i)
        const l2 = palindrome(s, i, i + 1)
        if (l1.length > res.length) res = l1;
        if (l2.length > res.length) res = l2;
    }
    return res
}

//leetcode submit region end(Prohibit modification and deletion)
